### PR TITLE
convert sync tests to go

### PR DESCRIPTION
**Pull request by charlierudolph**
Opening for visibility, not ready for review

### CODE REVIEW

**Comment by kevgo** [comment]
In case it helps, since there are so many sync tests, it might be more manageable to do one pull request per subdirectory.

File: test/steps/branch_steps.go [hunk 1]
```diff
@@ -93,12 +112,24 @@ func BranchSteps(suite *godog.Suite, fs *FeatureState) {
 		return fs.activeScenarioState.gitEnvironment.DeveloperRepo.CreateBranch(branch2, "main")
 	})
 
-	suite.Step(`^my repository has the feature branches "([^"]+)" and "([^"]+)"$`, func(branch1, branch2 string) error {
+	suite.Step(`^my repository has the (local )?feature branches "([^"]+)" and "([^"]+)"$`, func(localStr, branch1, branch2 string) error {
```
Comment (kevgo) [inline test/steps/branch_steps.go:115]: If you agree with the above, this needs updates as well.

File: test/steps/branch_steps.go [hunk 2]
```diff
@@ -109,16 +140,20 @@ func BranchSteps(suite *godog.Suite, fs *FeatureState) {
 		return fs.activeScenarioState.gitEnvironment.DeveloperRepo.PushBranch(branch1)
 	})
 
-	suite.Step(`^my repository has the perennial branches "([^"]+)" and "([^"]+)"$`, func(branch1, branch2 string) error {
+	suite.Step(`^my repository has the (local )?perennial branches "([^"]+)" and "([^"]+)"$`, func(localStr, branch1, branch2 string) error {
```
Comment (kevgo) [inline test/steps/branch_steps.go:143]: This one as well.

File: test/steps/coworker_steps.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"fmt"
+
+	"github.com/cucumber/godog"
+)
+
+// CoworkerSteps defines Gherkin step implementations around a coworker.
+func CoworkerSteps(suite *godog.Suite, fs *FeatureState) {
+	suite.Step(`^my coworker fetches updates$`, func() error {
+		_, err := fs.activeScenarioState.gitEnvironment.CoworkerRepo.Shell.Run("git", "fetch")
```
Comment (kevgo) [inline test/steps/coworker_steps.go:12]: We got tested and typesafe methods for such Git operations now. 🙂 

```go
_, err := fs.activeScenarioState.gitEnvironment.CoworkerRepo.Fetch()

File: test/steps/coworker_steps.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+		return err
+	})
+
+	suite.Step(`^my coworker sets the parent branch of "([^"]*)" as "([^"]*)"$`, func(childBranch, parentBranch string) error {
+		_, err := fs.activeScenarioState.gitEnvironment.CoworkerRepo.Shell.Run("git", "config", "git-town-branch."+childBranch+".parent", parentBranch)
```
Comment (kevgo) [inline test/steps/coworker_steps.go:17]: Such Git Town configuration operations are now centralized. You can use the [Configuration.SetParentBranch](https://github.com/git-town/git-town/blob/master/src/git/config.go#L411-L413) method here:

```go
_, err := fs.activeScenarioState.gitEnvironment.CoworkerRepo.Configuration(false).SetParentBranch(childBranch, parentBranch)

File: test/steps/tag_steps.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"fmt"
+
+	"github.com/cucumber/godog"
+	"github.com/cucumber/messages-go/v10"
+)
+
+// TagSteps defines Gherkin step implementations around merges.
+func TagSteps(suite *godog.Suite, fs *FeatureState) {
+	suite.Step(`^I have the following tags$`, func(table *messages.PickleStepArgument_PickleTable) error {
```
Comment (kevgo) [inline test/steps/tag_steps.go:12]: People don't have tags 🙂  Thoughts about renaming this to `my repo has the following tags`?

File: test/tag_table_bulider.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
+package test
```
Comment (kevgo) [inline test/tag_table_bulider.go:1]: Typo in filename `tag_table_bulider.go`

File: test/tag_table_bulider.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	//   tag2: [local, remote]
+	tagToLocations map[string]helpers.OrderedStringSet
+}
+
+// NewTagTableBuilder provides a fully initialized instance of TagTableBuilder.
+func NewTagTableBuilder() TagTableBuilder {
+	result := TagTableBuilder{
+		tagToLocations: make(map[string]helpers.OrderedStringSet),
+	}
+	return result
```
Comment (kevgo) [inline test/tag_table_bulider.go:27]: You could inline the `result` variable.

File: test/tag_table_bulider.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	if exists {
+		builder.tagToLocations[tag] = locations.Add(location)
+	} else {
+		builder.tagToLocations[tag] = helpers.NewOrderedStringSet(location)
+	}
+}
+
+// AddMany registers the given tags from the given location into this table.
+func (builder *TagTableBuilder) AddMany(tags []string, location string) {
+	for _, tag := range tags {
```
Comment (kevgo) [inline test/tag_table_bulider.go:42]: Consider more efficient loop here.

File: test/tag_table_bulider.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		builder.Add(tag, location)
+	}
+}
+
+// Table provides the data accumulated by this TagTableBuilder as a DataTable.
+func (builder *TagTableBuilder) Table() (result DataTable) {
+	result.AddRow("NAME", "LOCATION")
+	tags := make([]string, len(builder.tagToLocations))
```
Comment (kevgo) [inline test/tag_table_bulider.go:50]: Nice preallocation at the correct length!

File: test/commit_table_builder.go [hunk 10]
```diff
@@ -61,6 +61,13 @@ func (builder *CommitTableBuilder) Add(commit Commit, location string) {
 	}
 }
 
+// AddMany registers the given commits from the given location into this table.
+func (builder *CommitTableBuilder) AddMany(commits []Commit, location string) {
+	for _, commit := range commits {
```
[the remark this answers is not in the document]
Comment (charlierudolph) [reply test/commit_table_builder.go:66]: Makes sense. Hmm, would using an array of commit pointers be equivalent in efficiency? If so, I think I'd prefer we switch to that

File: test/git_environment.go [hunk 11]
```diff
@@ -34,29 +40,37 @@ func CloneGitEnvironment(original *GitEnvironment, dir string) (*GitEnvironment,
...
-	// Since we copied the files from the memoized directory,
-	// we have to set the "origin" remote to the copied origin repo here.
-	err = result.DeveloperRepo.AddRemote("origin", result.OriginRepo.Dir)
-	if err != nil {
-		return &result, fmt.Errorf("cannot set remote: %w", err)
-	}
-	err = result.DeveloperRepo.Fetch()
-	if err != nil {
-		return &result, fmt.Errorf("cannot fetch: %w", err)
+	for _, repo := range []GitRepository{result.DeveloperRepo, result.CoworkerRepo} {
```
[the remark this answers is not in the document]
Comment (charlierudolph) [reply test/git_environment.go:56]: Sounds good

File: test/git_repository.go [hunk 12]
```diff
@@ -279,6 +298,15 @@ func (repo *GitRepository) CurrentBranch() (result string, err error) {
 	return strings.TrimSpace(outcome.OutputSanitized()), nil
 }
 
+// CurrentFileContent provides the current content of a file.
+func (repo *GitRepository) CurrentFileContent(filename string) (result string, err error) {
```
[the remark this answers is not in the document]
Comment (charlierudolph) [reply test/git_repository.go:302]: Okay that works

File: test/git_repository.go [hunk 13]
```diff
@@ -315,6 +343,44 @@ func (repo *GitRepository) FilesInCommit(sha string) (result []string, err error
 	return strings.Split(strings.TrimSpace(outcome.OutputSanitized()), "\n"), nil
 }
 
+// FilesInBranch provides the list of the files present in the given branch.
+func (repo *GitRepository) FilesInBranch(branch string) (result []string, err error) {
+	outcome, err := repo.Shell.Run("git", "ls-tree", "-r", "--name-only", branch)
+	if err != nil {
+		return result, fmt.Errorf("cannot determine files in branch %q in repo %q: %w", branch, repo.Dir, err)
+	}
+	for _, line := range strings.Split(strings.TrimSpace(outcome.OutputSanitized()), "\n") {
```
[the remark this answers is not in the document]
Comment (charlierudolph) [reply test/git_repository.go:352]: Thanks for the cleanup!

File: test/steps/branch_steps.go [hunk 14]
```diff
@@ -73,16 +84,24 @@ func BranchSteps(suite *godog.Suite, fs *FeatureState) {
 		return fs.activeScenarioState.gitEnvironment.OriginRepo.CreateBranch(branch, "main")
 	})
 
-	suite.Step(`^my repository has a feature branch named "([^"]*)"$`, func(branch string) error {
-		return fs.activeScenarioState.gitEnvironment.DeveloperRepo.CreateFeatureBranch(branch)
+	suite.Step(`^my repository has a (local )?feature branch named "([^"]*)"$`, func(localStr, branch string) error {
```
[the remark this answers is not in the document]
Comment (charlierudolph) [reply test/steps/branch_steps.go:87]: Auhh, missed that distinction. I don't think the difference of "code base" vs "repository" is easy for someone to come in and understand. 

I think I'd prefer if we use "local" / "remote" qualifiers if talking about one of them and then we could use the plural form of repositories when referring to all? example "my repositories have the commits:"

File: test/steps/coworker_steps.go (hunk 3 above)
Comment (charlierudolph) [reply test/steps/coworker_steps.go:12]: Nice, will update

File: test/tag_table_bulider.go (hunk 9 above)
Comment (charlierudolph) [reply test/tag_table_bulider.go:50]: Linter made me do it haha

File: test/commit_table_builder.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (kevgo) [reply test/commit_table_builder.go:66]: I'd prefer to not switch to pointers for everything. They have other side effects like additional indirection on access (=slowness), and possibly more work for the garbage collector. Plus, on 64 bit systems pointers require an additional 8 bytes for every entry. 

I think it's better to use the loop it was designed in Go: with the loop index as the primary element, and using the optional copy of the current item only when that makes sense.

**Pull request merged**